core/staking_verifier.go

```diff
@@ -285,8 +305,34 @@ func VerifyAndDelegateFromMsg(
 			if startBalance.Cmp(delegateBalance) > 0 {
 				// Used undelegated token for redelegation
 				delegation.Undelegations = delegation.Undelegations[curIndex:]
+				shouldRemove := actualIndex != 0 && // never remove the (inactive) validator
+					len(delegation.Undelegations) == 0 &&
+					delegation.Amount.Cmp(common.Big0) == 0 &&
+					delegation.Reward.Cmp(common.Big0) == 0 &&
+					chainConfig.IsPostNoNilDelegations(epoch)
+				if shouldRemove {
+					modification := make(map[common.Address]uint64)
+					// mark for removal
+					modification[wrapper.Address] = math.MaxUint64
+					delegationsToAlter[delegation.DelegatorAddress] = modification
+					// mark the following ones with offset 1
+					for j := actualIndex + 1; j < uint64(len(wrapper.Delegations)); j++ {
```

Comment:
The continuous pruning seems is getting very complicated and also incurs extra cost by itself (such as this looping). And also, have you consider the case when more than 1 delegate/undelegate/collectreward txn happens in the same block? If the complicated is too high and the gain compared with batch pruning is not very high. We may not want to do the continuous pruning then.